dotnet/src/Microsoft.Agents.AI/ChatClient/ChatClientAgentStructuralOutput.cs

```diff
@@ -0,0 +1,197 @@
+    /// </para>
+    /// </remarks>
+    public async Task<AgentRunResponse<T>> RunAsync<T>(
+        IEnumerable<ChatMessage> messages,
+        AgentThread? thread = null,
+        JsonSerializerOptions? serializerOptions = null,
+        AgentRunOptions? options = null,
+        bool? useJsonSchemaResponseFormat = null,
+        CancellationToken cancellationToken = default)
+    {
+        var inputMessages = Throw.IfNull(messages) as IReadOnlyCollection<ChatMessage> ?? messages.ToList();
+
+        (ChatClientAgentThread safeThread, ChatOptions? chatOptions, List<ChatMessage> threadMessages) =
+            await this.PrepareThreadAndMessagesAsync(thread, inputMessages, options, cancellationToken).ConfigureAwait(false);
+
+        var chatClient = this.ChatClient;
+
+        chatClient = ApplyRunOptionsTransformations(options, chatClient);
+
+        var agentName = this.GetLoggingAgentName();
+
+        this._logger.LogAgentChatClientInvokingAgent(nameof(RunAsync), this.Id, agentName, this._chatClientType);
+
+        // Call the IChatClient and notify the AIContextProvider of any failures.
+        ChatResponse<T> chatResponse;
+        try
+        {
+            chatResponse = await chatClient.GetResponseAsync<T>(
+                threadMessages,
+                serializerOptions ?? AgentJsonUtilities.DefaultOptions,
+                chatOptions,
+                useJsonSchemaResponseFormat,
+                cancellationToken
+            ).ConfigureAwait(false);
+        }
+        catch (Exception ex)
+        {
+            await NotifyAIContextProviderOfFailureAsync(safeThread, ex, inputMessages, cancellationToken).ConfigureAwait(false);
+            throw;
+        }
+
+        this._logger.LogAgentChatClientInvokedAgent(nameof(RunAsync), this.Id, agentName, this._chatClientType, inputMessages.Count);
+
+        // We can derive the type of supported thread from whether we have a conversation id,
+        // so let's update it and set the conversation id for the service thread case.
+        this.UpdateThreadWithTypeAndConversationId(safeThread, chatResponse.ConversationId);
+
+        // Ensure that the author name is set for each message in the response.
+        foreach (ChatMessage chatResponseMessage in chatResponse.Messages)
+        {
+            chatResponseMessage.AuthorName ??= agentName;
+        }
+
+        // Only notify the thread of new messages if the chatResponse was successful to avoid inconsistent message state in the thread.
+        await NotifyThreadOfNewMessagesAsync(safeThread, inputMessages.Concat(chatResponse.Messages), cancellationToken).ConfigureAwait(false);
+
+        // Notify the AIContextProvider of all new messages.
+        await NotifyAIContextProviderOfSuccessAsync(safeThread, inputMessages, chatResponse.Messages, cancellationToken).ConfigureAwait(false);
+
+        return new AgentRunResponse<T>(chatResponse) { AgentId = this.Id };
```

Comment:
Do you need to duplicate all the code from RunAsync? Seems like the only things that differ are the GetResponseAsync call and the construction of the response at the end, in which case you could share most of it using a delegate or interface and two implementations or something like that?